From f3848d3ee06cc5b1924a35a9e3f207369658c9bd Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 27 Oct 2021 12:15:55 +0300
Subject: [PATCH] [SW-0] gaudi2: fix halt read status for arc

Currently as part of preboot where ARC0 is halted and reset
conditionally, the arc status is not read correctly instead register
offset is used to check status of ARC0. This patch add fix to read
register value using offset and check status of ARC0.

Change-Id: I3eb8f19c152c6e4cb83099f7442caf6e8baf619a
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 35e8751..bcdadb5 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -877,7 +877,7 @@
 	preboot_puts("\n");
 }
 
-u32 preboot_read_halt_status(u8 id)
+static u32 preboot_get_halt_offset(u8 id)
 {
 	u32 halt_reg_addr[] = {
 		mmPSOC_ARC0_CFG_RUN_HALT_ACK,
@@ -894,9 +894,11 @@
 
 void preboot_cond_arc0_reset(void)
 {
-	u32 halt_arc0_status = preboot_read_halt_status(0);
+	u32 halt_ack_offset = preboot_get_halt_offset(0);
+	u32 val;
 
-	if (halt_arc0_status & ARC_IS_RUNNING)
+	val = regmap_readl(halt_ack_offset);
+	if (val & ARC_IS_RUNNING)
 		return;
 
 	halt_n_reset_arc(0);
@@ -925,7 +927,7 @@
 		halt_ack_mask = PSOC_ARC0_CFG_RUN_HALT_ACK_HALT_ACK_MASK;
 	}
 
-	halt_ack_offset = preboot_read_halt_status(id);
+	halt_ack_offset = preboot_get_halt_offset(id);
 	val = regmap_readl(halt_ack_offset);
 	if (val & halt_mask) {
 		preboot_puts_hex("Skipping halt for ARC", id, false);
